### PR TITLE
Add ACME Plugin

### PR DESCRIPTION
**Pull request by chinzhiweiblank**
see [#3460](https://github.com/coredns/coredns/issues/3460)

### CODE REVIEW

**Comment by miekg** [comment]
Hi, this is neat. But before I do some reviewing, how does this interact with tls we use
in some plugins for supporting DoT and DoH ? Ideally I want to use this plugin to get them
certs and make stuff happening (ideally in a non-blocking [at startup] way).

Can you flesh out that scenario?

**Comment by chinzhiweiblank** [comment]
> Hi, this is neat. But before I do some reviewing, how does this interact with tls we use in some plugins for supporting DoT and DoH ? Ideally I want to use this plugin to get them certs and make stuff happening (ideally in a non-blocking [at startup] way). Can you flesh out that scenario?

It obtains the TLS configuration for the server like what the tls plugin does in [here](https://github.com/coredns/coredns/blob/5b9b079dabc7f71463cea3f0c6a92f338935039d/plugin/tls/tls.go#L87).

**Comment by yongtang** [comment]
@chinzhiweiblank Can you give an example Corefile setup?

**Comment by chrisohaver** [comment]
> Closes [#3011](https://github.com/coredns/coredns/pull/3011)

do you mean #3460 ?

**Comment by chinzhiweiblank** [comment]
> > Closes [#3011](https://github.com/coredns/coredns/pull/3011)
> 
> do you mean #3460 ?

My mistake. I changed it.

**Comment by chinzhiweiblank** [comment]
> @chinzhiweiblank Can you give an example Corefile setup?
@miekg 
If I understand your question correctly, I should be explaining how the keys in the certificate obtained in this plugin are used in DoT and DoH in other plugins.

```txt
acme {
  domain example.org
}
```

This will obtain the TLS certificate for `example.org`.
This plugin will do the following in this [segment](https://github.com/coredns/coredns/pull/4829/files#diff-58e97756fcb7e20904ac1abc9fe07f1a4d66e5d2b6401e2bd7e07ddce1df9d65R18-R24):
1. Fetch the configuration of the DNS server
2. Obtain the cached certificate
3. Assign the certificate to the configuration

**Comment by chrisohaver** [comment]
I think the question is "how does this interact with the _tls_ plugin ..."

**Comment by chrisohaver** [comment]
Missing: acme plugin added to `plugin.cfg`.

**Comment by chrisohaver** [comment]
> I think the question is "how does this interact with the _tls_ plugin ..."
> Missing: acme plugin added to plugin.cfg.

If placed after _tls_ in plugin.cfg: Looks like it would always override the _tls_ plugin.
If placed before _tls_ in plugin.cfg: incompatible with _tls_ causing CoreDNS not to start if both plugins are present.

**Comment by chrisohaver** [comment]
@chinzhiweiblank,  A while into reviewing this, I'm unsure of what the state of it is ...

Can you summarize the current state of this PR?  Is the plugin functional in its current state or is it a work in progress?  Is it ready for review?

**Comment by chinzhiweiblank** [comment]
> @chinzhiweiblank, A while into reviewing this, I'm unsure of what the state of it is ...
> 
> Can you summarize the current state of this PR? Is the plugin functional in its current state or is it a work in progress? Is it ready for review?

Hi Chris, I think it's a work in progress. It's not ready for review. Do you mind if I make it into a draft or close it for now? I would reopen it when I think it's ready.

**Comment by chinzhiweiblank** [comment]
> > I think the question is "how does this interact with the _tls_ plugin ..."
> > Missing: acme plugin added to plugin.cfg.
> 
> If placed after _tls_ in plugin.cfg: Looks like it would always override the _tls_ plugin.
> If placed before _tls_ in plugin.cfg: incompatible with _tls_ causing CoreDNS not to start if both plugins are present.

I would place it after tls. If the user chooses to use acme, he or she should be aware that the server would be using the keys obtained from acme in the end.

**Comment by chrisohaver** [comment]
> Hi Chris, I think it's a work in progress. It's not ready for review. Do you mind if I make it into a draft or close it for now? I would reopen it when I think it's ready.

Please mark it as a draft.

**Comment by chrisohaver** [comment]
Should this instead be a feature that extends the existing _tls_ plugin? I.e. new options in the _tls_ plugin.  I don't see any obvious _functional_ advantages to them being separate, since one would functionally eclipse the other if both are present.

**Comment by miekg** [comment]
[ Quoting ***@***.***> in "Re: [coredns/coredns] Add ACME Plug..." ]
>Should this instead be a feature that extends the existing tls plugin? I.e. new
>options in the tls plugin. I don't see any obvious functional advantages to
>them being separate, since one would functionally eclipse the other if both are
>present.

I think that makes sense, ie `tls acme { .... potential options here ....}`, and we need
to think throught the implications; are these certs stored on disk, what about storing
them in memcache? Should we explore caddy's Filesystem interface? This at minimum needs a
little design written down

**Comment by chrisohaver** [comment]
Stale, and I think preferable to extend the existing _tls_ plugin than to add a new plugin.
